File: backend/Modules/Candidature/Services/CandidatureService.cs
using NextStep.Modules.Candidature.DTOs;
using NextStep.Modules.Candidature.Repositories;
using NextStep.Shared.Pagination;
using CandidatureEntity = NextStep.Modules.Candidature.Models.Candidature;

namespace NextStep.Modules.Candidature.Services;

public class CandidatureService : ICandidatureService
{
    private readonly ICandidatureRepository _candidatureRepository;

    public CandidatureService(ICandidatureRepository candidatureRepository)
    {
        _candidatureRepository = candidatureRepository;
    }

    public async Task<CandidatureDto> CreateAsync(
        Guid userId,
        CreateCandidatureDto dto,
        CancellationToken cancellationToken = default)
    {
        var existing = await _candidatureRepository.GetByUserAndOfferAsync(
            userId,
            dto.IdOffre,
            cancellationToken);

        if (existing is not null)
        {
            return MapToDto(existing);
        }

        var entity = new CandidatureEntity
        {
            IdUtilisateur = userId,
            IdOffre = dto.IdOffre,
            InclureLettreMotivation = dto.InclureLettreMotivation,
            Statut = "EN_ATTENTE",
            DateCreation = DateTime.UtcNow
        };

        await _candidatureRepository.AddAsync(entity, cancellationToken);

        return MapToDto(entity);
    }

    public async Task<CandidatureDto?> GetByIdAsync(
        Guid candidatureId,
        CancellationToken cancellationToken = default)
    {
        var entity = await _candidatureRepository.GetByIdAsync(
            candidatureId,
            cancellationToken);

        return entity is null ? null : MapToDto(entity);
    }

    public async Task<List<CandidatureDto>> GetByUserIdAsync(
        Guid userId,
        CancellationToken cancellationToken = default)
    {
        var entities = await _candidatureRepository.GetByUserIdAsync(userId, cancellationToken);
        return entities.Select(MapToDto).ToList();
    }

    public async Task<PagedResponse<CandidatureDto>> GetByUserIdPagedAsync(
        Guid userId,
        int offset,
        int limit,
        bool interviewOnly = false,
        CancellationToken cancellationToken = default)
    {
        var safeOffset = Math.Max(0, offset);
        var safeLimit = Math.Clamp(limit, 1, 100);

        var (items, total) = await _candidatureRepository.GetByUserIdPagedAsync(
            userId,
            safeOffset,
            safeLimit,
            interviewOnly,
            cancellationToken);

        var mapped = items.Select(MapToDto).ToList();

        return new PagedResponse<CandidatureDto>
        {
            Offset = safeOffset,
            Limit = safeLimit,
            Total = total,
            HasMore = safeOffset + mapped.Count < total,
            Items = mapped
        };
    }

    private static CandidatureDto MapToDto(CandidatureEntity entity)
    {
        return new CandidatureDto
        {
            IdCandidature           = entity.IdCandidature,
            IdUtilisateur           = entity.IdUtilisateur,
            IdOffre                 = entity.IdOffre,
            DateCreation            = entity.DateCreation,
            InclureLettreMotivation = entity.InclureLettreMotivation,
            Statut                  = entity.Statut,
            ResponseStatus          = entity.ResponseStatus,
            HasResponse             = entity.HasResponse,
            LastCheckedAtUtc        = entity.LastCheckedAtUtc,
            LastResponseAtUtc       = entity.LastResponseAtUtc,
            // AI Classification
            LastResponseFrom        = entity.LastResponseFrom,
            LastResponseSnippet     = entity.LastResponseSnippet,
            ResponseSummary         = entity.ResponseSummary,
            RecommendedAction       = entity.RecommendedAction,
            ResponseConfidence      = entity.ResponseConfidence,
            ResponseClassifiedAtUtc = entity.ResponseClassifiedAtUtc,
            // Follow-up tracking
            FollowUpNeeded          = entity.FollowUpNeeded,
            LastFollowUpAtUtc       = entity.LastFollowUpAtUtc,
        };
    }
}

File: frontend/src/app/services/candidature.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, catchError, map } from 'rxjs';
import { environment } from '../../environments/environment';

// ── DTOs matching backend ────────────────────────────────────────────────────

export interface CandidatureDto {
  idCandidature: string;
  idUtilisateur: string;
  idOffre: string;
  dateCreation: string;
  inclureLettreMotivation: boolean;
  statut: string;
  responseStatus: string;
  hasResponse: boolean;
  lastCheckedAtUtc?: string;
  lastResponseAtUtc?: string;

  // ── AI Classification ─────────────────────────────────────────────────────────
  lastResponseFrom?: string;
  lastResponseSnippet?: string;
  responseSummary?: string;
  recommendedAction?: string;
  responseConfidence?: number;
  responseClassifiedAtUtc?: string;

  // ── Follow-up tracking (set by Hangfire DetectFollowUpNeededJob) ──────────────
  followUpNeeded?: boolean;
  lastFollowUpAtUtc?: string;
}

export interface CreateCandidaturePayload {
  idOffre: string;
  inclureLettreMotivation?: boolean;
}

export interface PagedResponse<T> {
  offset: number;
  limit: number;
  total: number;
  hasMore: boolean;
  items: T[];
}

// ── Service ──────────────────────────────────────────────────────────────────

@Injectable({ providedIn: 'root' })
export class CandidatureService {
  private readonly http = inject(HttpClient);
  private readonly base = environment.apiBaseUrl;

  getMyCandidatures(): Observable<CandidatureDto[]> {
    return this.http.get<CandidatureDto[]>(`${this.base}/candidatures`);
  }

  getMyCandidaturesPaged(offset = 0, limit = 10, interviewOnly = false): Observable<PagedResponse<CandidatureDto>> {
    return this.http
      .get<PagedResponse<CandidatureDto>>(
        `${this.base}/candidatures/paged?offset=${offset}&limit=${limit}&interviewOnly=${interviewOnly}`
      )
      .pipe(
        catchError(() =>
          this.getMyCandidatures().pipe(
            map((all) => {
              const filtered = interviewOnly
                ? all.filter((c) => {
                    const status = `${c.statut ?? ''} ${c.responseStatus ?? ''}`.toUpperCase();
                    return status.includes('ENTRETIEN');
                  })
                : all;

              const sorted = [...filtered].sort(
                (a, b) =>
                  new Date(b.lastResponseAtUtc ?? b.dateCreation).getTime() -
                  new Date(a.lastResponseAtUtc ?? a.dateCreation).getTime()
              );

              const safeOffset = Math.max(0, offset);
              const safeLimit = Math.max(1, limit);
              const items = sorted.slice(safeOffset, safeOffset + safeLimit);
              const total = sorted.length;

              return {
                offset: safeOffset,
                limit: safeLimit,
                total,
                hasMore: safeOffset + items.length < total,
                items,
              } as PagedResponse<CandidatureDto>;
            })
          )
        )
      );
  }

  getById(id: string): Observable<CandidatureDto> {
    return this.http.get<CandidatureDto>(`${this.base}/candidatures/${id}`);
  }

  create(payload: CreateCandidaturePayload): Observable<CandidatureDto> {
    return this.http.post<CandidatureDto>(`${this.base}/candidatures`, payload);
  }
}
